File: device/cuda/matmul/native/matmul.c
#include "matmul.h"
#include "../internal/bridge/core_private.h"

#include <stdio.h>

static const char* g_cuda_matmul_module_source = NULL;

void cuda_matmul_register_module_source(const char* source) {
    g_cuda_matmul_module_source = source;
}

const char* cuda_matmul_module_source(void) {
    return g_cuda_matmul_module_source;
}

void cuda_matmul_status_clear(CUDAStatus* status) {
    cuda_status_clear(status);
}

void cuda_matmul_status_set(CUDAStatus* status, int code, const char* message) {
    cuda_status_set(status, code, message);
}

int cuda_matmul_kernel_name(
    char* out,
    size_t outBytes,
    const char* operationName,
    int elementDType,
    CUDAStatus* status
) {
    const char* suffix = cuda_element_dtype_suffix(elementDType);

    if (suffix == NULL) {
        cuda_matmul_status_set(status, -6, "unknown CUDA matmul dtype");
        return -6;
    }

    return cuda_compose_kernel_name(out, outBytes, operationName, suffix, status);
}

int cuda_matmul_dispatch_tiled(
    CUDADeviceRef contextRef,
    const char* operationName,
    int elementDType,
    CUDABufferRef leftRef,
    CUDABufferRef rightRef,
    CUDABufferRef biasRef,
    CUDABufferRef outRef,
    uint32_t rows,
    uint32_t inner,
    uint32_t cols,
    int hasBias,
    uint64_t completionToken,
    CUDAStatus* status
) {
    cuda_matmul_status_clear(status);

    if (rows == 0 || inner == 0 || cols == 0) {
        return 0;
    }

    if (leftRef == NULL || rightRef == NULL || outRef == NULL) {
        cuda_matmul_status_set(status, -2, "nil CUDA matmul buffer");
        return -2;
    }

    if (hasBias && biasRef == NULL) {
        cuda_matmul_status_set(status, -2, "nil CUDA matmul bias buffer");
        return -2;
    }

    char kernelName[128];
    int nameCode = cuda_matmul_kernel_name(
        kernelName,
        sizeof(kernelName),
        operationName,
        elementDType,
        status
    );

    if (nameCode != 0) {
        return nameCode;
    }

    const char* moduleSource = cuda_matmul_module_source();

    if (moduleSource == NULL) {
        cuda_matmul_status_set(status, -7, "CUDA matmul module source not registered");
        return -7;
    }

    CUDAContext* context = NULL;
    CUDAStreamRef stream = NULL;
    int prepareCode = cuda_context_prepare(contextRef, status, &context, &stream);

    if (prepareCode != 0) {
        return prepareCode;
    }

    CUDAKernelRef kernel = cuda_get_kernel(context, moduleSource, kernelName, status);

    if (kernel == NULL) {
        return status != NULL && status->code != 0 ? status->code : -7;
    }

    void* leftPtr = cuda_buffer_device_ptr(leftRef);
    void* rightPtr = cuda_buffer_device_ptr(rightRef);
    void* outPtr = cuda_buffer_device_ptr(outRef);
    void* biasPtr = hasBias ? cuda_buffer_device_ptr(biasRef) : NULL;

    uint32_t gridX = (cols + 15u) / 16u;
    uint32_t gridY = (rows + 15u) / 16u;
    uint32_t sharedBytes = 256u * 2u * (uint32_t)sizeof(float);

    if (hasBias) {
        void* args[] = {&leftPtr, &rightPtr, &biasPtr, &outPtr, &rows, &inner, &cols};
        int launchCode = cuda_launch_grid(
            context,
            kernel,
            stream,
            gridX,
            gridY,
            1,
            16,
            16,
            1,
            sharedBytes,
            args,
            sizeof(args),
            status
        );

        if (launchCode != 0) {
            return launchCode;
        }
    }

    if (!hasBias) {
        void* args[] = {&leftPtr, &rightPtr, &outPtr, &rows, &inner, &cols};
        int launchCode = cuda_launch_grid(
            context,
            kernel,
            stream,
            gridX,
            gridY,
            1,
            16,
            16,
            1,
            sharedBytes,
            args,
            sizeof(args),
            status
        );

        if (launchCode != 0) {
            return launchCode;
        }
    }

    cuda_track_completion(context, stream, completionToken, NULL, status);
    return 0;
}

File: device/metal/matmul/native/matmul.m
#include "matmul.h"

#include "../internal/bridge/core_private.h"

#include <Foundation/Foundation.h>
#include <Metal/Metal.h>
#import <MetalPerformanceShaders/MetalPerformanceShaders.h>
#include <stdio.h>

void metal_matmul_status_clear(MetalStatus* status) {
    if (status == NULL) {
        return;
    }

    status->code = 0;
    status->message[0] = '\0';
}

void metal_matmul_status_set(MetalStatus* status, int code, const char* message) {
    if (status == NULL) {
        return;
    }

    status->code = code;

    if (message == NULL) {
        status->message[0] = '\0';
        return;
    }

    snprintf(status->message, METAL_STATUS_MESSAGE_BYTES, "%s", message);
}

static const char* metal_matmul_dtype_suffix(int elementDType) {
    switch (elementDType) {
    case MetalElementDTypeFloat32: return "float32";
    case MetalElementDTypeFloat16: return "float16";
    case MetalElementDTypeBFloat16: return "bfloat16";
    default: return NULL;
    }
}

int metal_matmul_kernel_name(
    char* out,
    size_t outBytes,
    const char* operationName,
    int elementDType,
    MetalStatus* status
) {
    const char* suffix = metal_matmul_dtype_suffix(elementDType);

    if (operationName == NULL || suffix == NULL) {
        metal_matmul_status_set(status, -6, "unknown Metal matmul kernel");
        return -6;
    }

    int written = snprintf(out, outBytes, "%s_%s", operationName, suffix);

    if (written <= 0 || (size_t)written >= outBytes) {
        metal_matmul_status_set(status, -6, "Metal matmul kernel name overflow");
        return -6;
    }

    return 0;
}

static int metal_matmul_mps_element_size(int elementDType, MetalStatus* status) {
    switch (elementDType) {
    case MetalElementDTypeFloat32: return (int)sizeof(float);
    case MetalElementDTypeFloat16: return (int)sizeof(uint16_t);
    case MetalElementDTypeBFloat16: return (int)sizeof(uint16_t);
    default:
        metal_matmul_status_set(status, -6, "unsupported Metal matmul dtype");
        return 0;
    }
}

static MPSDataType metal_matmul_mps_dtype(int elementDType, MetalStatus* status) {
    switch (elementDType) {
    case MetalElementDTypeFloat32: return MPSDataTypeFloat32;
    case MetalElementDTypeFloat16: return MPSDataTypeFloat16;
    case MetalElementDTypeBFloat16: return MPSDataTypeBFloat16;
    default:
        metal_matmul_status_set(status, -6, "unsupported MPS matmul dtype");
        return MPSDataTypeInvalid;
    }
}

static bool metal_matmul_mps_row_bytes_aligned(uint32_t columnCount, int elementSize) {
    NSUInteger rowBytes = (NSUInteger)columnCount * (NSUInteger)elementSize;

    return rowBytes % 256 == 0;
}

static int metal_matmul_dispatch_mps(
    MetalDeviceRef contextRef,
    int elementDType,
    MetalBufferRef leftRef,
    MetalBufferRef rightRef,
    MetalBufferRef outRef,
    uint32_t rows,
    uint32_t inner,
    uint32_t cols,
    uint64_t completionToken,
    MetalStatus* status
) {
    @autoreleasepool {
        metal_matmul_status_clear(status);

        MetalContext* context = (MetalContext*)contextRef;

        if (context == NULL || context->device == NULL || context->queue == NULL) {
            metal_matmul_status_set(status, -1, "invalid Metal context");
            return -1;
        }

        int elementSize = metal_matmul_mps_element_size(elementDType, status);

        if (elementSize == 0) {
            return status != NULL && status->code != 0 ? status->code : -6;
        }

        if (!metal_matmul_mps_row_bytes_aligned(inner, elementSize)
            || !metal_matmul_mps_row_bytes_aligned(cols, elementSize)) {
            return -100;
        }

        MPSDataType dataType = metal_matmul_mps_dtype(elementDType, status);

        if (dataType == MPSDataTypeInvalid) {
            return status != NULL && status->code != 0 ? status->code : -6;
        }

        id<MTLDevice> device = (__bridge id<MTLDevice>)context->device;
        MPSMatrixDescriptor* leftDescriptor = [MPSMatrixDescriptor
            matrixDescriptorWithRows:rows
            columns:inner
            rowBytes:inner * (NSUInteger)elementSize
            dataType:dataType
        ];
        MPSMatrixDescriptor* rightDescriptor = [MPSMatrixDescriptor
            matrixDescriptorWithRows:inner
            columns:cols
            rowBytes:cols * (NSUInteger)elementSize
            dataType:dataType
        ];
        MPSMatrixDescriptor* outDescriptor = [MPSMatrixDescriptor
            matrixDescriptorWithRows:rows
            columns:cols
            rowBytes:cols * (NSUInteger)elementSize
            dataType:dataType
        ];

        MPSMatrix* leftMatrix = [[MPSMatrix alloc]
            initWithBuffer:(__bridge id<MTLBuffer>)leftRef
            descriptor:leftDescriptor
        ];
        MPSMatrix* rightMatrix = [[MPSMatrix alloc]
            initWithBuffer:(__bridge id<MTLBuffer>)rightRef
            descriptor:rightDescriptor
        ];
        MPSMatrix* outMatrix = [[MPSMatrix alloc]
            initWithBuffer:(__bridge id<MTLBuffer>)outRef
            descriptor:outDescriptor
        ];

        MPSMatrixMultiplication* matmul = [[MPSMatrixMultiplication alloc]
            initWithDevice:device
            transposeLeft:NO
            transposeRight:NO
            resultRows:rows
            resultColumns:cols
            interiorColumns:inner
            alpha:1.0
            beta:0.0
        ];

        id<MTLCommandQueue> queue = (__bridge id<MTLCommandQueue>)context->queue;
        id<MTLCommandBuffer> commandBuffer;

        if (context->isBatching) {
            if (context->currentCommandBuffer == NULL) {
                id<MTLCommandBuffer> batchBuffer = [queue commandBuffer];
                context->currentCommandBuffer = (__bridge_retained void*)batchBuffer;
            }

            commandBuffer = (__bridge id<MTLCommandBuffer>)context->currentCommandBuffer;
            metal_suspend_compute_encoder(context);
        } else {
            commandBuffer = [queue commandBuffer];
        }

        [matmul encodeToCommandBuffer:commandBuffer leftMatrix:leftMatrix rightMatrix:rightMatrix resultMatrix:outMatrix];
        metal_track_command_completion(context, commandBuffer, completionToken, NULL);

        if (!context->isBatching) {
            [commandBuffer commit];
        }

        return 0;
    }
}

static int metal_matmul_dispatch(
    MetalDeviceRef contextRef,
    const char* kernelName,
    uint32_t rows,
    uint32_t cols,
    uint64_t completionToken,
    MetalStatus* status,
    void (^encode)(id<MTLComputeCommandEncoder> encoder)
) {
    @autoreleasepool {
        metal_matmul_status_clear(status);

        MetalContext* context = (MetalContext*)contextRef;

        if (context == NULL || context->queue == NULL) {
            metal_matmul_status_set(status, -1, "invalid Metal context");
            return -1;
        }

        id<MTLComputePipelineState> pipeline = metal_get_pipeline(context, kernelName, status);

        if (pipeline == nil) {
            return status != NULL && status->code != 0 ? status->code : -7;
        }

        id<MTLCommandQueue> queue = (__bridge id<MTLCommandQueue>)context->queue;
        id<MTLCommandBuffer> commandBuffer;
        id<MTLComputeCommandEncoder> encoder = metal_get_encoder((MetalContext*)contextRef, &commandBuffer);

        [encoder setComputePipelineState:pipeline];
        encode(encoder);
        [encoder
            dispatchThreadgroups:MTLSizeMake((cols + 15) / 16, (rows + 15) / 16, 1)
            threadsPerThreadgroup:MTLSizeMake(16, 16, 1)
        ];
        metal_track_command_completion((MetalContext*)contextRef, commandBuffer, completionToken, NULL);
        metal_end_encoder((MetalContext*)contextRef, encoder, commandBuffer);

        return 0;
    }
}

